### PR TITLE
[fsdp] feat: add NPU fusion kernels for Qwen3 MoE

### PR DESCRIPTION
**Pull request by Shangwei-Li**
### What does this PR do?

This PR adds following  NPU fusion kernels to Qwen3 MoE model in Transformers: GroupedMatMul, SwiGLU, RMSNorm and RoPE.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/volcengine/verl/pulls?q=fusion++npu+moe
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

Tested with Qwen3-30B-A3B sp8 fsdp32 on Ascend A2:
Without kernel fusion:
image
WIth kernel fusion
image

Test results with train_prompt_bsz=512 sp8 fsdp32 on Ascend A2. The  orange line represents GPU, the pink line represents NPU, max absolute error in reward is less than 5%.
image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/models/transformers/npu_patch.py [hunk 1]
```diff
@@ -46,5 +50,125 @@ def rms_norm_forward(self, x):
     return torch_npu.npu_rms_norm(x, self.weight, epsilon=self.variance_epsilon)[0]
 
 
+def apply_rotary_pos_emb_qwen3_npu(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (tardis-key) [inline verl/models/transformers/npu_patch.py:50]: this funcdoes not seem to be limited to 'Qwen3' scenario. It is recommended to remove 'qwen3' from its name to facilitate reuse by other developers.

File: verl/models/transformers/npu_patch.py [hunk 2]
```diff
@@ -46,5 +50,125 @@ def rms_norm_forward(self, x):
     return torch_npu.npu_rms_norm(x, self.weight, epsilon=self.variance_epsilon)[0]
 
 
+def apply_rotary_pos_emb_qwen3_npu(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = torch_npu.npu_rotary_mul(q, cos, sin)
```
Comment (FightingZhen) [inline verl/models/transformers/npu_patch.py:53]: So curious if this implementation is strictly equal to the api `apply_rotary_emb` in `flash-attn`, because I have tried similar implementation in [transformers](https://github.com/huggingface/transformers/pull/38491/files#diff-e54265e770cee1df09627645e6629dbc3675dda3a33abab63635652fd67ceafcR253)

File: verl/models/transformers/npu_patch.py [hunk 3]
```diff
@@ -46,5 +50,125 @@ def rms_norm_forward(self, x):
...
 
+def apply_rotary_pos_emb_qwen3_npu(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = torch_npu.npu_rotary_mul(q, cos, sin)
+    k_embed = torch_npu.npu_rotary_mul(k, cos, sin)
+    return q_embed.to(q.dtype), k_embed.to(k.dtype)
+
+
+class torch_npu_gmm(torch.autograd.Function):
```
Comment (FightingZhen) [inline verl/models/transformers/npu_patch.py]: use lower-case name as class name is strange, suggest defining normal class name, and define specific func name below it.

File: verl/models/transformers/npu_patch.py [hunk 4]
```diff
@@ -46,5 +50,125 @@ def rms_norm_forward(self, x):
...
+    @staticmethod
+    def forward(ctx, x, weight, group_list, split_size):
+        ctx.save_for_backward(x, weight)
+        ctx.group_list = group_list
+        ctx.split_size = split_size
+
+        outputs = torch_npu.npu_grouped_matmul([x], [weight], group_list=group_list, group_type=0, split_item=2)
```
Comment (FightingZhen) [inline verl/models/transformers/npu_patch.py:65]: is input param `x` and `weight` with `[]` required?

File: verl/models/transformers/npu_patch.py [hunk 5]
```diff
@@ -46,5 +50,125 @@ def rms_norm_forward(self, x):
...
+        x, weight = ctx.saved_tensors
+        group_list = ctx.group_list
+        wt = weight.permute(0, 2, 1)
+        xt = x.permute(1, 0)
+        dx = torch_npu.npu_grouped_matmul([grad_outputs], [wt], group_list=group_list, group_type=0, split_item=2)
+        dw = torch.zeros_like(weight)
+        split_size = ctx.split_size
+        xt_list = torch.split(xt, split_size, dim=1)
+        grad_outputs_list = torch.split(grad_outputs, split_size, dim=0)
+        with torch.npu.amp.autocast(enabled=False):
```
Comment (FightingZhen) [inline verl/models/transformers/npu_patch.py:79]: if `enabled=False`, can we remove it?

File: verl/models/transformers/npu_patch.py [hunk 6]
```diff
@@ -46,5 +50,125 @@ def rms_norm_forward(self, x):
...
+            dw = torch.stack([torch.matmul(xt_list[i], grad_outputs_list[i]) for i in range(len(xt_list))])
+
+        return dx[0], dw, None, None
+
+
+class Qwen3MoeSparseMoeBlockPatch(nn.Module):
+    def __init__(self, config):
```
Comment (FightingZhen) [inline verl/models/transformers/npu_patch.py]: `__init__` func seems as same as original code in `transformers==4.52.4`, can we remove it?

File: verl/models/transformers/npu_patch.py (hunk 1 above)
Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py:50]: This is because the signature of functioin `apply_rotary_pos_emb` in qwen3 is diffirent from qwen2.5 in transformer. This subfix is added to avoid confusion.

File: verl/models/transformers/npu_patch.py (hunk 4 above)
Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py:65]: Yes, `torch_npu.npu_grouped_matmul` API requires list of tensors as input.

File: verl/models/transformers/npu_patch.py (hunk 5 above)
Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py:79]: AMP is not enabled with FSDP backend. But as a general torch function, bfloat16 gmm backward result will be different from `mindspeed.ops.gmm`(which has been used and tested for a long time) since mindspeed ops is not affected by AMP and large matmul is very sensitive to dtype.

File: verl/models/transformers/npu_patch.py (hunk 6 above)
Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py]: Removed.

File: verl/models/transformers/npu_patch.py (hunk 3 above)
Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py]: Fixed.

File: verl/models/transformers/npu_patch.py (hunk 2 above)
Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py:53]: According to the current transformers code, `qwen3_moe.py` does not call the `apply_rotary_pos_emb` function in `flash-attn`. 
Furthermore, We have compared the results of fused rope, original transformers rope on GPU/NPU, with reuslts of  float32 rope on CPU as true values. In terms of cosine similarity and eular distance, reuslts of fused NPU rope are closer to true values than those of original GPU/NPU implementation.
image

File: verl/models/transformers/npu_patch.py [hunk 7]
```diff
@@ -16,16 +16,17 @@
...
 from transformers.models.qwen2_5_vl import modeling_qwen2_5_vl
-from transformers.models.qwen2_5_vl.modeling_qwen2_5_vl import Qwen2RMSNorm
+from transformers.models.qwen3_moe import modeling_qwen3_moe
 
 
 # This patch takes effect when using apply_rotary_pos_emb_flashatt on qwen2_5_vl and will be removed in
 # subsequent versions
 # https://github.com/huggingface/transformers/pull/38491
-def apply_rotary_pos_emb_flashatt_npu(
+def apply_rotary_pos_emb_flashatt_qwen2vl_npu(
```
Comment (FightingZhen) [inline verl/models/transformers/npu_patch.py]: 这个是不是应该是qwen2_5_vl？

Comment (Shangwei-Li) [reply verl/models/transformers/npu_patch.py]: Fixed.

**Comment by FightingZhen** [comment]
LGTM, can we review and merge it? @vermouth1992

**Review by vermouth1992 — APPROVED** [review]
